### ISSUE TITLE
Cannot load a simple model save by PyTorch

### ISSUE THREAD

**Issue by stereomatchingkiss** [opencv/opencv#9501]
os : ubuntu16.04.1
opencv : 3.3
compiler : gcc5.6
python : 3.5

I train and save a very simple cnn model by PyTorch
```
class conv_block(nn.Module):
    def __init__(self, in_filter, out_filter, kernel):
        super(conv_block, self).__init__()

        self.conv1 = nn.Conv2d(in_filter, out_filter, kernel, 1, (kernel - 1)//2)
        self.batchnorm = nn.BatchNorm2d(out_filter)
        self.maxpool = nn.MaxPool2d(2, 2)

    def forward(self, x):
        x = self.conv1(x)
        x = self.batchnorm(x)
        x = F.relu(x)
        x = self.maxpool(x)

        return x

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()

        self.conv1 = conv_block(3, 6, 3)
        self.conv2 = conv_block(6, 16, 3)
        self.fc1 = nn.Linear(16 * 8 * 8, 120)
        self.bn1 = nn.BatchNorm1d(120)
        self.fc2 = nn.Linear(120, 84)
        self.bn2 = nn.BatchNorm1d(84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = x.view(x.size()[0], -1)
        x = F.relu(self.bn1(self.fc1(x)))
        x = F.relu(self.bn2(self.fc2(x)))
        x = self.fc3(x)
        return x

```

I save it by two solutions

```
torch.save(net.state_dict(), 'cifar10_model')
torch.save(net, 'cifar10_model.net')
```

Then load it by c++ api

```
std::string const model_file("/home/some_folder/cifar10_model");

std::cout<<"read net from torch"<<std::endl;
dnn::Net net = dnn::readNetFromTorch(model_file);
```

It always throw exception

**OpenCV Error: The function/feature is not implemented (Unsupported Lua type) in readObject, file /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/torch/torch_importer.cpp, line 797 /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/torch/torch_importer.cpp:797: error: (-213) Unsupported Lua type in function readObject**

But the module only use Conv2d, ReLU, BatchNorm2d, MaxPool2d and Linear layer only, every layers are supported by opencv3.3, ENet of torch is more complicated than this one. I am doing something wrong or this is a bug of 3.3?

**Comment by stereomatchingkiss** [comment]
resnet18 model download from [official site](https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py) throw the same error messages too, pretty sure this is a bug/unfinished features of 3.3

**Comment by dkurt** [comment]
Hi, @stereomatchingkiss! Currently DNN supports models from Caffe, TensorFlow and Torch frameworks. There are a python wrapper for Torch that is called as pytorch: https://github.com/hughperkins/pytorch and a one more PyTorch (http://pytorch.org/) that looks like a separate project. Some of its developers are the same, though.

**Comment by stereomatchingkiss** [comment]
Hi, @dkurt , thanks for you info, but I think your answer is not related to the issue. 

I know PyTorch(the codes I posted is written by PyTorch), my question is "why a simple network describe by PyTorch cannot load by dnn module", the answer is dnn do not support nngraph yet.

**Comment by stereomatchingkiss** [comment]
@dkurt I restudy your answer, do you mean opencv do not support PyTorch([http://pytorch.org/](http://pytorch.org/)) by pytorch([https://github.com/hughperkins/pytorch](https://github.com/hughperkins/pytorch))?If that is the case, please specify it clearly on the document, it is confusing, I believe today most of the people only think PyTorch  is ([https://github.com/pytorch/pytorch](https://github.com/pytorch/pytorch)) but not pytorch([https://github.com/hughperkins/pytorch](https://github.com/hughperkins/pytorch)).

In this case, although I do not like tensorflow(horrible api) nor caffe(clumsy engineer products, apparently it is designed by non-engineer people), I think I have to use it if I want my model portable.

**Comment by stereomatchingkiss** [comment]
By the way, any plan to support PyTorch(http://pytorch.org/)?

**Comment by dkurt** [comment]
@stereomatchingkiss, thank you for the note! May you try to define a model using `nn`-like way? Using `nn.Sequential` and add layers by `add()` method?

**Comment by stereomatchingkiss** [comment]
@dkurt I give [PyTorch](http://pytorch.org/) nn.Sequential a shot, they give me the same errors.

Model definition:

```
class Flatten(nn.Module):
    def forward(self, x):
        return x.view(-1, 16*5*5)

model = nn.Sequential(
                nn.Conv2d(3,6,5),
                nn.ReLU(),                
                nn.MaxPool2d(2, 2),
                nn.Conv2d(6, 16, 5),
                nn.ReLU(),
                nn.MaxPool2d(2, 2),
                Flatten(),                
                nn.Linear(16 * 5 * 5, 120),
                nn.ReLU(),
                nn.Linear(120, 84),
                nn.ReLU(),
                nn.Linear(84, 10)
                )

```

Save as

```
torch.save(model.state_dict(), "model_0.tch")
torch.save(model, "model_1.tch")

```

Both of them do not work, I guess I would stick to tensorflow(always avoid it because I think their api are very horrible). Would opencv support [PyTorch](http://pytorch.org/) in the future? Thanks

**Comment by dkurt** [comment]
@stereomatchingkiss, I think `nn.View` or `nn.Reshape` layer instead a function insertion would help.

**Comment by stereomatchingkiss** [comment]
@dkurt [PyTorch ](http://pytorch.org/docs/master/genindex.html)do not have nn.Reshape nor nn.View, you need to call the view function from Tensor.

**Comment by l-bat** [comment]
You can export PyTorch model to ONNX file and run ONNX model in OpenCV #12071

**Comment by dkurt** [comment]
@l-bat, it looks like with merged https://github.com/opencv/opencv/pull/12071 we can manage to run this network and receive comparable outputs with PyTorch? If so, we can close an issue.

**Comment by zjchuyp** [comment]
@l-bat, thank you for your ONNX module, but I found it don't support ONNX model containing LSTM layer.

**Comment by dkurt** [comment]
@zjchuyp, feel free to open a new feature request including a way to create a model with LSTM (at least an origin framework). OpenCV supports LSTM layer and this is mostly parser's issue.

**Comment by prasadpant** [comment]
Will cv2.dnn.readNetFromTorch feature available with official version PyTorch(http://pytorch.org/) without having to convert the PyTorch to ONNX module ?

**Comment by MohamedHesham02** [comment]
how to convert my pytorch custom model to onnx ?
